### PR TITLE
optimize Dynarray.of_list

### PR DESCRIPTION
**Pull request by gasche**
I missed a natural optimization of Dynarray.of_list, which is to use List.length to guess the right size for the resulting dynarray instead of using the dynamic resizing strategy.

A microbenchmark suggests that this is always a win, small (10% speedup) on small lists and large (2x-3x faster) on large lists.

### CODE REVIEW

**Review by yallop — APPROVED** [review]
Clearly correct.

(There still seems like quite a lot of unnecessary overhead in the implementation, e.g. from the branches in `ensure_capacity` and `add_last`, and the updating of `a.length` for every list element.)

**Comment by gasche** [comment]
I got nerd-sniped by your comment and wrote an implementation in the spirit of `of_array`. I cannot observe a significant performance difference in my benchmarks. I think that the branches get predicted well, and that the extra integer write is dominated by the caml_modify inherent in each addition. An unboxed implementation of dynarray (spoiler alert, I might be working on this) would let us do better here.

**Comment by gasche** [comment]
(I am not sure whether this PR needs two maintainer approvals because it is in the stdlib, or whether one suffices because it does not touch the API. In doubt I will just wait.)

**Review by nojb — APPROVED** [review]
LGTM

**Comment by gasche** [comment]
Thanks! Merging.

**Pull request merged**
